docs: create Sentry Design document (#903)
platform-platform/monorepo: metrics/web/docs/features/sentry/01_sentry_design.md

Example of the overriding the `FlutterError.onError` property:

```dart
FlutterError.onError = (details, {bool forceReport = false}) {
    Zone.current.handleUncaughtError(details.exception, details.stack);
};
```

### Capturing exceptions

As described in the [previous step](#capturing-uncaught-flutter-and-dart-errors), we have an ability to catch uncaught Flutter and Dart errors, but we still need to catch specific errors in the try-catch block. To do that we have to use a Sentry SDK `captureException` method.

Example: 

```dart
try {
  aMethodThatMightFail();
} catch (exception, stackTrace) {
  await Sentry.captureException(
    exception,
    stackTrace: stackTrace,
  );
}
```

### Adding contexts

The Contexts interface provides an additional context data. For example, the device or browser version. The contexts type can be used to define arbitrary contextual data on the event. To do that we have to use a Sentry SDK `configureScope` method.

To add a new custom context, we should give it a name and related key/value pairs:

```dart
final projectName = {'name' : 'test'};
Sentry.configureScope((scope) => scope.setContexts('project', projectName));
```

> Also, the Sentry SDK contains specific classes to facilitate adding a [common context](https://develop.sentry.dev/sdk/event-payloads/contexts/):
> - To create an app context, we should use an `App` class.
> - To create a browser context, we should use a `Browser` class.
> - To create a device context, we should use a `Device` class.
> - To create an operation system context, we should use an `OperationSystem` class.
> - To create a GPU context, we should use a `GPU` class.
> - To create a runtime context, we should use a `Runtime` class.

Reviewer: ```suggestion
The Sentry SDK provides several specific classes to facilitate adding a [common context](https://develop.sentry.dev/sdk/event-payloads/contexts/):
- `App` class with application-related information.
- `Browser` class with browser-related information.
// and so on.
- To create a device context, we should use a `Device` class.
- To create an operation system context, we should use an `OperationSystem` class.
- To create a GPU context, we should use a `GPU` class.
- To create a runtime context, we should use a `Runtime` class.
```

> Example of adding a common context using the Sentry SDK built-in classes:

Reviewer: ```suggestion
Consider an example of using the `Browser` class to set the browser common context:
```

```dart
final browser = Browser(name: 'chrome', version: '87.0.4280.88');
Sentry.configureScope((scope) => scope.setContexts(Browser.type, browser));
```

## Sentry Util

> To provide an easy way to use the Sentry SDK, we should create a wrapper named `SentryUtil`. 
> 
> The wrapper should contain methods, that meet our requirements described in the [design steps](#design):

Reviewer: ```suggestion
To simplify working with Sentry SDK, we are to implement a utility class that would wrap Sentry calling. This wrapper has to contain methods that meet the requirements described in the [Design section](#design). Here is a list of these methods:
```

> - a `configureDefaultContexts` private method that configures default contextual data on the error event. It should add context about the current user `operation system`, and depending on running the application on the `web` or `mobile` - about the current user `browser` or `device`;

Reviewer: ```suggestion
- a `configureDefaultContexts` method that configures default contextual data on the error event. This method adds all common contexts as user's `operation system`, `browser` information, and so on;
```

> - an `initialize` method , that simplify [initializing the Sentry](#initializing-the-sentry) with an ability to [capture uncaught Dart and Flutter errors](#capturing-uncaught-flutter-and-dart-errors), set the DSN and release name;

Reviewer: ```suggestion
- an `initialize` method that simplifies [initializing the Sentry](#initializing-the-sentry) process;
```

> - a `captureException` method, that delegates to the `Sentry` method described in the [Capturing exceptions](#capturing-exceptions);

Reviewer: ```suggestion
- a `captureException` method that delegates to the `Sentry` capture exceptions;
```

> - an `addContext` method, that delegates to the `Sentry` method described in the [Adding Contexts](#adding-contexts).

Reviewer: ```suggestion
- an `addContext` method that simplifies adding contexts to the `Sentry`;
```

> The following class diagram represents the classes required for Sentry integration: 

Reviewer: ```suggestion
The following class diagram describes the structure of Sentry integration: 
```

![Sentry class diagram](http://www.plantuml.com/plantuml/proxy?cache=no&fmt=svg&src=https://github.com/platform-platform/monorepo/raw/sentry_design/metrics/web/docs/features/sentry/diagrams/sentry_util_class_diagram.puml)

> ## Data Source Name (DSN)

Reviewer: ```suggestion
## Sentry Options Binding
```

> 

Reviewer: ```suggestion

The following subsections describe how to bind `DSN` and `release` options to the Sentry SDK using your build environment.

### Data Source Name (DSN)
```

> The [DSN](https://docs.sentry.io/product/sentry-basics/dsn-explainer/) tells the SDK where to send the events. If this value is not provided, the SDK will try to read it from the `SENTRY_DSN` environment variable. If that variable also does not exist, the SDK will just not send any events.
> 
> The `DSN` is not a secret, worst thing someone could do is sending events to your account. If that ever happens we have a few options tackling this, we can either [block off certain request](https://docs.sentry.io/product/accounts/quotas/#inbound-data-filters) or cycle the `DSN`.
> 
> Despite the above, we still should to improve our security in using the `DSN`, so we should build the Flutter app providing the `DNS` enviroment variable. Also, the environment variable save us from the hard-coding the `DSN`.

Reviewer: ```suggestion
The [DSN](https://docs.sentry.io/product/sentry-basics/dsn-explainer/) tells the SDK where to send the events. If this value is not provided, the SDK will try to read it from the `SENTRY_DSN` environment variable. If that variable also does not exist, the SDK will just not send any events.

The `DSN` can be public as it does not contain any secret data. Despite this fact, someone can use it to send events to your project. There are a few options to tackle this:
- block off certain requests using [inbound data filters](https://docs.sentry.io/product/accounts/quotas/#inbound-data-filters);
- secure your `DSN`.

Let's focus on how to secure your `DNS` and prevent someone from using it. The main idea is to use the `SENTRY_DSN` environment variable while building your app. Consider the following example:
```

```bash
export SENTRY_DSN=`your_dsn_here`
flutter build web --dart-define=SENTRY_DSN=$SENTRY_DSN
```

> Then in the app code we can take the `DSN` using `String.fromEnviroment()` method:

Reviewer: ```suggestion
Then within the application, you can access the variable value using the [`String.fromEnviroment`](link-goes-here) method:
```

```dart

> const dsn = const String.fromEnvironment('SENTRY_DSN');

Reviewer: ```suggestion
const sentryDsn = String.fromEnvironment('SENTRY_DSN');
```
https://dart.dev/guides/language/effective-dart/usage#dont-use-const-redundantly

```

> 

Reviewer: ```suggestion

### Release

```
I suppose we should provide and information on how we can set the release as well. Consider the following points to cover in this sub-section:
- describe why we need the release to be set;
- notice that for the Android and IOS the Sentry SDK takes the release automatically (https://docs.sentry.io/platforms/flutter/configuration/releases/#bind-the-version); 
- provide examples in what can be a release value for the web application;
- describe defining a variable and using it within the application.

Consider the link: https://docs.sentry.io/platforms/flutter/configuration/releases/

> ## Updating source maps

Reviewer: ```suggestion
## Updating Source Maps
```

> Due to flutter minifying, the release app errors are not readable in the Sentry. To resolve the problem we should [update the js sourcemaps](https://docs.sentry.io/platforms/javascript/sourcemaps). 

Reviewer: ```suggestion
Flutter minifies JavaScript code building the web application. This makes applications faster but also results in not readable errors and their stack traces. To resolve this problem we should [update the JS sourcemaps](https://docs.sentry.io/platforms/javascript/sourcemaps) each time we build a new application release. 
```

> To get it work in the `Dart` we should build the Flutter app providing the `release name` and updating the source maps:

Reviewer: ```suggestion
Updating source maps requires the [`Sentry CLI`](https://docs.sentry.io/product/cli/) to be installed. Also, it requires the [release binding](...) described in the previous section. Consider the following example:
```

```bash
export SENTRY_RELEASE=release-name
flutter build web --dart-define=SENTRY_RELEASE=$SENTRY_RELEASE

# Upload source maps
sentry-cli releases new $SENTRY_RELEASE

sentry-cli releases files $SENTRY_RELEASE upload-sourcemaps . \
    --ext dart \
    --rewrite

cd ./build/web/
sentry-cli releases files $SENTRY_RELEASE upload-sourcemaps . \
    --ext map \
    --ext js \
    --rewrite

sentry-cli releases finalize $SENTRY_RELEASE
```

Then in the app code, we should take the `release name` using `String.fromEnviroment()` method and pass it to the `SentryUtil.init` method:

```dart
const release = const String.fromEnvironment('SENTRY_RELEASE');

> ```

Reviewer: Should be in the section with release binding.

> 

Reviewer: ```suggestion

## Testing

```
I suppose we should add information on how we can test Sentry logging. Consider the following points to cover: 
- note the sentry has the `.bindClient` method (https://pub.dev/documentation/sentry/latest/sentry/Sentry/bindClient.html);
- note that using this method we can use the mockito to test sentry;
- note that as sentry is a third party API it possible should be tested using [third party API testing](https://github.com/platform-platform/monorepo/blob/master/docs/03_third_party_api_testing.md) with [mock server](https://github.com/platform-platform/monorepo/blob/master/docs/04_mock_server.md).

Not sure we will test Sentry integration widely but we have to describe the possible testing.

```dart
const release = const String.fromEnvironment('SENTRY_RELEASE');
```

## Updating Source Maps

Flutter minifies JavaScript code building the web application. This makes applications faster but also results in not readable errors and their stack traces. To resolve this problem we should [update the JS sourcemaps](https://docs.sentry.io/platforms/javascript/sourcemaps) each time we build a new application release.

Updating source maps requires the [`Sentry CLI`](https://docs.sentry.io/product/cli/) to be installed. Also, it requires the [release binding](#release) described in the previous section. Consider the following example:

```bash
export SENTRY_RELEASE=release-name
flutter build web --dart-define=SENTRY_RELEASE=$SENTRY_RELEASE

# Upload source maps
sentry-cli releases new $SENTRY_RELEASE

sentry-cli releases files $SENTRY_RELEASE upload-sourcemaps . \
    --ext dart \
    --rewrite

cd ./build/web/
sentry-cli releases files $SENTRY_RELEASE upload-sourcemaps . \
    --ext map \
    --ext js \
    --rewrite

sentry-cli releases finalize $SENTRY_RELEASE
```

## Testing

> The Sentry should be tested using a [mockito](https://pub.dev/packages/mockito) package, which allows replacing a [`SentryClient`](https://pub.dev/documentation/sentry/latest/sentry/SentryClient-class.html) with a mock one using a [`Sentry.bindClient`](https://pub.dev/documentation/sentry/latest/sentry/Sentry/bindClient.html) method. 
> Also, the approaches discussed in [3rd-party API testing](https://github.com/platform-platform/monorepo/blob/master/docs/03_third_party_api_testing.md) and [here](https://github.com/platform-platform/monorepo/blob/master/docs/04_mock_server.md) should be used testing a Sentry client direct HTTP calls. 

Reviewer: ```suggestion
The main idea in testing Sentry integration is to use the [`Sentry.bindClient`](https://pub.dev/documentation/sentry/latest/sentry/Sentry/bindClient.html) method and bind a test client to use in test cases. 

As Sentry is a third-party integration it should be tested using the [third-party API testing](https://github.com/platform-platform/monorepo/blob/master/docs/03_third_party_api_testing.md) approach that uses the [mock server](https://github.com/platform-platform/monorepo/blob/master/docs/04_mock_server.md). Also, it is possible to mock the client using a [mockito](https://pub.dev/packages/mockito) package.
```
